woocommerce/wc-admin
# Add `/reports/customers` endpoint

Fixes #916.

This PR seeks to implement the `/reports/customers` endpoint.

It introduces a new lookup table, `wc_customer_lookup`. Schema discussion: p7bje6-1dM-p2.

### Detailed test instructions:

- Run `phpunit` and make sure all tests are passing.
- Ensure that your test store has some registered customers
- Deactivate/activate `wc-admin` to kick off customer lookup table creation
- Verify that all registered customers have a row in `wc_customer_lookup`
- Verify that checking out as a guest results in a new row in `wc_customer_lookup` without a `user_id`
- Verify that registering at checkout results in a new row in `wc_customer_lookup` with a `user_id`
- Use Postman or another tool to make some test requests to `/wc/v3/reports/customers`
Available params:
`page`
`per_page`
`order` (`asc`, `desc`)
`orderby` (`username`, `name`, `country`, `city`, `postcode`, `date_registered`, `date_last_active`, `orders_count`, `total_spend`, `avg_order_value`)
`match` (`any`, `all`)
`name`
`username`
`email`
`country`
`last_active_before`
`last_active_after`
`registered_before`
`registered_after`
`orders_count_min`
`orders_count_max`
`total_spend_min`
`total_spend_max`
`avg_order_value_min`
`avg_order_value_max`
`last_order_before`
`last_order_after`

Note that all "range" type filters will be `AND`ed regardless of `match` value. E.g. `orders_count_min = 5, orders_count_max = 10` always results in `(orders_count >= 5 AND orders_count <= 10)`.
